lib/hooks.js

```diff
@@ -119,6 +119,67 @@ exports.addHook = function(name, hooks, stage, type, inlineConfig){
 };
 
 /**
+ * Merge multiple hookset collected from resources, customTypes, etc
+ * @param hookSet - 
+ */
+exports.merge = function(hookset, rec) {
+  var isArray = _.any(hookset, function(arg) { return _.isArray(arg); });
+  if(isArray) {
+    return _.reduce(hookset, function(result, value, key) {
+      return result.concat(value || []);
+    }, []);
+  }
+
+  var keys = _.unique(_.flatten(_.map(hookset, function(arg) {
+    return _.keys(arg);
+  })));
+
+  return _.object(_.map(keys, function(key) {
+    return [key, exports.merge(
+      _.compact(_.map(hookset, function(arg) {
+        return arg[key];
+      })), 1
+    )];
+  }));
+}
+
+
+/* Collect hooks provided with custom types
+ * and convert then to a resource-wide hooks
+ * @param customTypes - a map with field names => custom types for particular resource
+ */
+exports.fromCustomTypesMap = function(customTypes) {
+  return _.flatten(_.map(_.keys(customTypes), function(key) {
+    // Modify a names of the hooks to include field name they applied to
+    var result = _.cloneDeep(customTypes[key].hooks);
+    _.each(result, function(whenHooks, when) {
+      _.each(whenHooks, function(actionHooks, action) {
+        _.each(actionHooks, function(ahook) {
+          if(ahook.name) {
+            ahook.name = [key, ahook.name].join('-');
+          }
+          // Set highest priority
+          ahook.priority = 1000;
```

Comment:
It's not guaranteed to be highest here :) Pretty sure there're use-cases where it's set to 1000000. Can we read existing hooks priorities and take `max + 1` ? Not sure if all hooks are set up by the rime it runs.